cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -72,6 +72,371 @@ constexpr inline auto is_supported_representation_type()
+ * @return Returns value / 10^exp10
+ */
+template <typename T>
+CUDF_HOST_DEVICE inline T divide_power10_32bit(T value, int exp10)
+{
+  // Computing division this way is much faster than the alternatives.
+  // Division is not implemented in GPU hardware, and the compiler will often implement it as a
+  // multiplication of the reciprocal of the denominator, requiring a conversion to floating point.
+  // Ths is especially slow for larger divides that have to use the FP64 pipeline, where threads
+  // bottleneck.
+
+  // Instead, if the compiler can see exactly what number it is dividing by, it can
+  // produce much more optimal assembly, doing bit shifting, multiplies by a constant, etc.
+  // For the compiler to see the value though, array lookup (with exp10 as the index)
+  // is not sufficient: We have to use a switch statement. Although this introduces a branch,
+  // it is still much faster than doing the divide any other way.
+  // Perhaps an array can be used in C++23 with the assume attribute?
+
+  // Since we're optimizing division this way, we have to do this for multiplication as well.
+  // That's because doing them in different ways (switch, array, runtime-computation, etc.)
+  // increases the register pressure on all kernels that use fixed_point types, specifically slowing
+  // down some of the PYMOD and join benchmarks.
+
+  // This is split up into separate functions for 32-, 64-, and 128-bit denominators.
+  // That way we limit the templated, inlined code generation to the exponents that are
+  // capable of being represented. Combining them together into a single function again
+  // introduces too much pressure on the kernels that use this code, slowing down their benchmarks.
+
+  // Also note that these divisors are hardcoded here in place, rather than in the PowersOf10
+  // struct. This is because the NotEqual benchmark is slower when doing that; I guess the compiler
+  // is getting confused by all of the templating and inlining and can't optimize it as well?
+
+  // Note: Using signed powers of 10 (where possible) to avoid unintended integer conversion
+
+  switch (exp10) {
+    case 0: return value;
+    case 1: return value / 10;
+    case 2: return value / 100;
+    case 3: return value / 1000;
+    case 4: return value / 10000;
+    case 5: return value / 100000;
+    case 6: return value / 1000000;
+    case 7: return value / 10000000;
+    case 8: return value / 100000000;
+    case 9: return value / 1000000000;
+    default: return 0;
+  }
+}
+
+/**
+ * @brief Divide by a power of 10 that fits within a 64bit integer.
+ *
+ * @tparam T Type of value to be divided-from.
+ * @param value The number to be divided-from.
+ * @param exp10 The power-of-10 of the denominator, from 0 to 19 inclusive.
+ * @return Returns value / 10^exp10
+ */
+template <typename T>
+CUDF_HOST_DEVICE inline T divide_power10_64bit(T value, int exp10)
+{
+  // See comments in divide_power10_32bit() for discussion.
+  // clang-format off
```

Comment:
I'd prefer to leave clang-format on.

Reply:
Turning it off makes it line-up as:

    case 0: return value;
    case 1: return value / 10;
    case 2: return value / 100;
    case 3: return value / 1000;
    case 4: return value / 10000;
    case 5: return value / 100000;
    case 6: return value / 1000000;
    case 7: return value / 10000000;
    case 8: return value / 100000000;
    case 9: return value / 1000000000;
    case 10: return value / 10000000000LL;
    case 11: return value / 100000000000LL;
    case 12: return value / 1000000000000LL;
    case 13: return value / 10000000000000LL;
    case 14: return value / 100000000000000LL;
    case 15: return value / 1000000000000000LL;
    case 16: return value / 10000000000000000LL;
    case 17: return value / 100000000000000000LL;
    case 18: return value / 1000000000000000000LL;
    case 19: return value / 10000000000000000000ULL;  // 10^19 only fits if unsigned!
    default: return 0;

Reply:
I don’t see the problem. Let’s turn it off.

Reply:
With clang-format off it's much easier to see that everything is increasing by a power of 10 when they're all lined up at the same position.  It's not that important to me though, so I'll re-enable clang-format here.